.github/workflows/continous-integration-workflow.yml

```diff
@@ -25,4 +26,53 @@ jobs:
       uses: rokroskar/workflow-run-cleanup-action@master
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-    if: "!startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master'"
\ No newline at end of file
+    if: "!startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master'"
+
+  # Production build
+  productionbuild:
+    name: Production Build and test
+    runs-on: ubuntu-latest      # Sets the build environment a machine with the latest Ubuntu installed
+    steps:
+    - name: Checkout the code    # Adds a step to checkout the repository code
+      uses: actions/checkout@v2 
+    - name: Build the application
+      run: |
+        docker build --target production --tag lakshmips/todo-app-service .
+    - name: Run the unit and integration tests
+      working-directory: .
+      run: | 
+        docker build --target test --tag todo-app:test .
+        docker run --env-file .env.test todo-app:test  
+    - name: Cleanup action
+      uses: rokroskar/workflow-run-cleanup-action@master
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    if: "!startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master'"
+
+  dockerhubpublish:
+    name: Build and push to Docker Hub
+    runs-on: ubuntu-latest      # Sets the build environment a machine with the latest Ubuntu installed
+    defaults:
+      run:
+        working-directory: .
+    needs: productionbuild
```

Comment:
Should we also rely on the build stage (or is that meant to be relied on by `productionbuild`)?
Otherwise the tests could fail there but still allow this to push & continue - note that the display shown in the Actions page shows no connections between that build stage and anything else!

Reply:
This is where I was confused, I have just relied on Productionbuild here as I have the lines...
docker build --target test --tag todo-app:test .
 docker run --env-file .env.test todo-app:test  
These lines will run the tests and will fail the build if there is a error.

What is the best practice in here...
Should the productionbuild be relying on build, so that I can avoid running the tests in productionbuild. Could you assist please

Reply:
It's a good question - relying on both will slow your pipeline down, but  currently
* You run the unit tests twice - but we only pay attention to the second run. So flaky/inconsistent tests could trigger a concern but still let the pipeline pass which is confusing if nothing else
* You only run the end-to-end tests once, and you don't fail the build if they fail!
Personally I'd lean towards having productionbuild depend on build _or_ having dockerhubpublish depend on both - potentially allowing them to still run at the same time; or potentially removing the productionbuild stage entirely (if it's just building it to check the image can be built, then is there any reason not to just do that at the end of build?)
If you do keep productionbuild and go with adding dependencies, then we could always remove the repeat testing step